lib/created_message_queue_handler.ts

```diff
@@ -223,56 +287,77 @@ export async function handleMessage(
-        }`
-      )
+    .alt(defaultAddresses.chain(getEmailAddressFromDefaultAddresses))
+    .foldL<Promise<Either<RuntimeError, Option<NotificationEvent>>>>(
+      // if we didn't get any destination address, just log a warning
+      async () => {
+        winston.warn(
+          `Fiscal code has no associated email address and no default email address was provided|${
+            newMessageWithContent.fiscalCode
+          }`
+        );
+        return right(none);
+      },
+      // if we got a destination address from profile or defaults,
+      // try to create an email notification to enqueue
+      async notificationChannelEmail =>
+        await createNotification(
+          lNotificationModel,
+          senderMetadata,
+          newMessageWithContent,
+          NotificationChannelEnum.EMAIL,
+          notificationChannelEmail
+        )
     );
+
+  if (isLeft(errorOrMaybeEmailNotification)) {
+    return left(errorOrMaybeEmailNotification.value);
   }
+  const maybeEmailNotification = errorOrMaybeEmailNotification.value;
 
-  const newNotification = createNewNotification(
-    ulidGenerator,
-    newMessageWithContent.fiscalCode,
-    newMessageWithContent.id
-  );
+  //
+  //  Webhook notification
+  //
 
-  const errorOrNotification = await lNotificationModel.create(
-    {
-      ...newNotification,
-      channels: {
-        [NotificationChannelEnum.EMAIL]: maybeNotificationChannelEmail.toUndefined()
-      }
-    },
-    newNotification.messageId
+  // whether the recipient wants us to send notifications to the app backend
+  const isWebhookEnabled = maybeProfile.exists(
+    profile => profile.isWebhookEnabled === true
   );
 
-  if (isLeft(errorOrNotification)) {
-    return left(TransientError("Cannot save notification to database"));
-  }
-
-  const notification = errorOrNotification.value;
+  const errorOrMaybeWebhookNotification = isWebhookEnabled
+    ? await createNotification(
+        lNotificationModel,
+        senderMetadata,
+        newMessageWithContent,
+        NotificationChannelEnum.WEBHOOK,
+        { url: lDefaultWebhookUrl }
+      )
+    : right<RuntimeError, Option<NotificationEvent>>(none);
 
-  const emailNotification: NotificationEvent = {
-    message: {
-      ...newMessageWithContent,
-      kind: "INewMessageWithContent"
-    },
-    notificationId: notification.id,
-    senderMetadata
+  if (isLeft(errorOrMaybeWebhookNotification)) {
+    return left(errorOrMaybeWebhookNotification.value);
+  }
+  const maybeWebhookNotification = errorOrMaybeWebhookNotification.value;
+
+  //
+  //  Return notification events (one for each channel)
+  //
+  const outputBindings: OutputBindings = {
+    emailNotification: maybeEmailNotification.toUndefined(),
+    webhookNotification: maybeWebhookNotification.toUndefined()
   };
 
-  return right({ emailNotification });
+  // avoid to enqueue messages for non existing notifications
+  return right(withoutUndefinedValues(outputBindings));
```

Comment:
instead of transforming the `Option`s to `.toUndefined()` and then calling `withoutUndefinedValues` you could skip the `.toUndefined()` and filter with `isSome`

Reply:
not really clear, may you provide an example ?

Reply:
sorry I always forget there's not filter on maps in ts, never mind - what I meant was basically having a method like `withoutUndefinedValues` that filtered out empty `Option`s